### ISSUE TITLE
Feature requests

### ISSUE THREAD

**Issue by jsmitman82** [tomba/netserializer#20]
1) Don't require Initialize to be called on all types at program start, instead during serialize/deserialize do a dictionary lookup to see if the type has been generated before, and if not then generate that one type.

2) Don't error out if [Serialize] isn't on a class, simply skip it.  If I don't want a base class to be serialized, I'd like the option of not marking it as such, and having the serializer skip it.

3) Add a Deserialize() that uses ref instead of out, or some other means of deserializing to an object that has already been allocated.  Don't force allocation of a new object.

**Comment by itadapter** [comment]
The [1] above is doable, at least in general, I have added this to our serializer that has improved speed by good 50+%. Now we can process 120-150K 2-way net calls/sec instead of 70K.

https://github.com/aumcode/nfx/blob/master/Source/NFX/Serialization/Slim/SlimSerializer.cs#L112-L134

We emit the type handle which is like "$1" whee the int is the number of the already know type handle, the "BatchMode" accumulates the already transmitted types.

Use case of "stateful channel" that remembers what types have been written:

https://github.com/aumcode/nfx/blob/master/Source/Testing/NUnit/NFX.NUnit/Serialization/SlimBatchTypeMode.cs

**Comment by tomba** [comment]
The [1] is about a new feature, not optimization. And I cannot see how [1] could be faster than the current switch-case.

So yes, I agree it's doable. But it would require considerable plumbing and somehow dynamically sending type information back and forth between the two ends. Maybe somebody will show me wrong, but it feels to me much slower and much more complex than what NetSerializer does now. So I should rephrase my "not doable" to "not doable sensibly".

And while it may be comparing apples and oranges as I have no idea what your speed test does, on my laptop with NetSerializer I get 3.5M simple messages per second (one way) through a tcp socket (send and receive on same machine).

**Comment by itadapter** [comment]
Firstly, lets start from the fact that I am an author of the similar lib. What I suggest to you is based only on PRACTICAL use-cases, not a "what-if" out of nowhere.

Take a case of 5-6 generics base types closed by 50+ actual types (a typical example of business logic)
this would create 5*50 actual types( Agreement<T>, Cancelation<T>..., Service, Contract, Agreement, Client,....etc...). It is neither possible nor desired to know up-front what types will be marshaled. Also, this is a VERY IMPORTANT feature for custom heap managers (that you have probably not considered at all). The memory manager does not know what types will be written into byte[], hence you will need to emit full type spec which can be huge, especially for generics.
The adaptive "type registry" allows us to synchronize client and server - that is the stream gets auto-compressed as more and more types are synthesized. 

Say you need to keep sending: PurchaseContract<Customer, Dog>, but then you need to send PurchaseContract<Company, Elephant>. And there are hundreds of combinations.

As for 3.5 Mil/sec you are not comparing apples to apples at all. I was talking about WCF-like interface-based high-level service framework, not tcp socket. I do push 500K high-level ONE-WAY messages a second (that is with inspection filtering and authentication) between 2 machines on 1 Gb network. 

https://github.com/aumcode/nfx/tree/master/Source/NFX/Glue

**Comment by tomba** [comment]
Ok, let me rephrase again: "not doable sensibly for NetSerializer, and its targeted use cases" =).

Yes, I had the hunch that we are talking about rather different use cases. "WCF-like interface-based high-level service framework" is rather far from the target for NetSerializer.

I'm still wondering what you meant with your original comment of improving speed. Are you saying that something here could somehow improve the speed of NetSerializer, or that it improved the speed of your serializer? I'm always interested in improving the speed of NetSerializer, but I'm not interested in reducing the performance by making it more complex, as there are a lot of more complex serializers out there.

**Comment by itadapter** [comment]
There is going to be a release within a few weeks of
https://github.com/aumcode/serbench

Your serializer is already included as one of the represented technologies (10+ others).
We will be able to see how well different serializer supports cases when 500 unknown types need to be sent over and deserialized back a few thousand times. This will be  fair comparison.
You will then be able to get the answer to the question you are asking me.
All I can say - from preliminary results, that 90% of all serializers can not pass most simple tests 
(like  IMyData[,]  or ref cycle or struct { readonly field} or Tuple....

**Comment by tomba** [comment]
Ok, that will be interesting. Although "500 unknown types" or cyclical refs are very much something NetSerializer was not designed to handle. And that was by purpose, to keep it simple and as fast as possible.

But as long as the benchmark also covers simpler scenarios, it'll be interesting to see the results. If it only covers those high level dynamic scenarios you were describing, well, I'm sure it is still interesting for many people but I could already tell those people not to use NetSerializer =).

**Comment by itadapter** [comment]
The handling of cyclical references adds 50% overhead to CPU at best, if not 100%. The protobuf-like serializers that majority of users are so crazy about have major problem - they are not "transparent", as they require decoration with attributes/IDL files. They also can not map to common CLR types (even things like structs) - which makes it a great pain in the butt when all peers in your solution are .NET controlled by you, yet you need to create crappy DTO types just to be able to do a RPC call.  From what I understood, you built your serializers for that very purpose - to transparantly move data between NET machines. 
Ours does the same thing, only we support full set of features (references, multidim arrays etc, ISerializable etc.) which comes at cost of like 25% slow down compared to direct graph writing without references. However, the payoff is huge, as you can now transmit anything you want (i.e. ConcurrentDictionary<struct, Dictionary<.....>>).

As for SerbBech tests we are testing things like "TypicalPerson" with 8 fields. Then List of TypicalPerson, then array of lists, then array of structs with back refs etc...

**Comment by tomba** [comment]
I hear you, attributes/IDL files are annoying =). That said, I do see their value in many use cases. Just not in my use case.

To be honest, I built NetSerializer really for one purpose only, to do serialization for my Dwarrowdelf game. I later made it a separate library for others to use as it didn't require much work from my side, and I thought is should work fine for may simpler use cases other people might have.

In Dwarrowdelf's case I never transmit "live" data, that is, structs/classes that are used by the application. Instead I always create a new message, and fill the relevant bits and pieces there. For example, the server might send information to the client about a sword. But it cannot send the sword object as such, as it contains lots of data the client should not know about. So the server creates a new message and fills in only the pieces that the client can know about.

As almost all communication is similar to the above case, I never need to transmit classes like ConcurrentDictionary, nor do I have cyclical references, and I know all the message types and the types that the messages can contain.

I probably should improve the README to describe better the intended use cases, as every now and then someone asks questions about NetSerializer and it's revealed that they are trying to use it for something it's not designed for.
